### PR TITLE
feat(gccjit): Allow set_loccation to various classes

### PR DESCRIPTION
**Pull request by tempdragon**
Including:
1. field
2. function
3. rvalue

This is part of my initial draft of debuginfo. It is part of the prerequisites to merge my debuginfo support. 
Note: Only C interface is supported and only RValue is currently in use. Of course, if you insist, I can also create a commit with other extra parts removed.

### CODE REVIEW

**Comment by tempdragon** [comment]
See also: rust-lang/rustc_codegen_gcc#69

File: gcc/jit/jit-recording.cc [hunk 1]
```diff
@@ -5188,6 +5188,21 @@ recording::block::get_loc () const
...
+void
+recording::block::set_loc (location * loc)
+{
+  recording::statement *stmt =
+      (m_statements.length ())?
+      stmt = m_statements[0]:
+      NULL;
+
+  if (stmt)
+    return stmt->set_loc (loc);
```
Comment (antoyo) [inline gcc/jit/jit-recording.cc:5203]: I would prefer something like this:
```suggestion
  if (m_statements.length () > 0)
      m_statements[0]->set_loc (loc);
```

File: gcc/jit/jit-recording.cc [hunk 2]
```diff
@@ -5188,6 +5188,21 @@ recording::block::get_loc () const
     return NULL;
 }
 
+/* Set the source-location of a block by using that of the first
+   statement within it, if any.  */
+
+void
+recording::block::set_loc (location * loc)
```
Comment (antoyo) [inline gcc/jit/jit-recording.cc:5195]: Does it have any effect on the generated code to set the location of a block like this? I would think that setting the locations on the statements directly should be enough, no?

File: gcc/jit/libgccjit.cc [hunk 3]
```diff
@@ -1974,7 +1974,7 @@ gcc_jit_rvalue_as_object (gcc_jit_rvalue *rvalue)
   RETURN_NULL_IF_FAIL (rvalue, NULL, NULL, "NULL rvalue");
 
   return static_cast <gcc_jit_object *> (rvalue->as_object ());
-}
+}
```
Comment (antoyo) [inline gcc/jit/libgccjit.cc:1977]: Please remove this change.

File: gcc/jit/libgccjit.cc [hunk 4]
```diff
@@ -4749,3 +4749,59 @@ gcc_jit_context_add_top_level_asm (gcc_jit_context *ctxt,
...
+
+/* Public entrypoint.  See description in libgccjit.h.
+
+   After error-checking, this calls the trivial
+   gcc::jit::recording::field::set_loc method, in jit-recording.h.  */
+
+void
+gcc_jit_field_set_location (gcc_jit_field *field, gcc_jit_location *loc)
+{
+  RETURN_IF_FAIL (field, NULL, NULL, "NULL array type");
```
Comment (antoyo) [inline gcc/jit/libgccjit.cc:4761]:
```suggestion
  RETURN_IF_FAIL (field, NULL, NULL, "NULL field");
```

File: gcc/jit/libgccjit.cc [hunk 5]
```diff
@@ -4749,3 +4749,59 @@ gcc_jit_context_add_top_level_asm (gcc_jit_context *ctxt,
...
+/* Public entrypoint.  See description in libgccjit.h.
+
+   After error-checking, this calls the trivial
+   gcc::jit::recording::function::set_loc method, in jit-recording.h.  */
+
+void
+gcc_jit_function_set_location (gcc_jit_function *func,
+			     gcc_jit_location *loc)
+{
+  RETURN_IF_FAIL (func, NULL, NULL, "NULL rvalue");
```
Comment (antoyo) [inline gcc/jit/libgccjit.cc:4790]:
```suggestion
  RETURN_IF_FAIL (func, NULL, NULL, "NULL func");
```

File: gcc/jit/libgccjit.cc [hunk 6]
```diff
@@ -4749,3 +4749,59 @@ gcc_jit_context_add_top_level_asm (gcc_jit_context *ctxt,
...
+/* Public entrypoint.  See description in libgccjit.h.
+
+   After error-checking, this calls the trivial
+   gcc::jit::recording::block::set_loc method, in jit-recording.h.  */
+
+void
+gcc_jit_block_set_location (gcc_jit_block *blk,
+			     gcc_jit_location *loc)
+{
+  RETURN_IF_FAIL (blk, NULL, NULL, "NULL rvalue");
```
Comment (antoyo) [inline gcc/jit/libgccjit.cc:4804]:
```suggestion
  RETURN_IF_FAIL (blk, NULL, NULL, "NULL blk");
```

File: gcc/jit/libgccjit.map [hunk 7]
```diff
@@ -340,3 +339,10 @@ LIBGCCJIT_ABI_36 {
   global:
     gcc_jit_function_new_temp;
 } LIBGCCJIT_ABI_35;
+LIBGCCJIT_ABI_37 {
```
Comment (antoyo) [inline gcc/jit/libgccjit.map:342]: Please add an empty line before this line.

File: gcc/jit/libgccjit.map [hunk 8]
```diff
@@ -105,7 +105,6 @@ LIBGCCJIT_ABI_0
     gcc_jit_type_get_const;
     gcc_jit_type_get_pointer;
     gcc_jit_type_get_volatile;
-
```
Comment (antoyo) [inline gcc/jit/libgccjit.map:108]: Please remove this change.

**Comment by tempdragon** [comment]
I will also update the change to gccjit.rs later.

File: gcc/jit/jit-recording.cc (hunk 2 above)
Comment (tempdragon) [reply gcc/jit/jit-recording.cc:5195]: I was actually uncertain about whether to add such a `set_loc` for gccjit or to simply allow setting locations for all statements if possible. When the code was written, I was unsure about the future.
In the current version only rvalue is used. Maybe I should simply remove all other code.

File: gcc/jit/libgccjit.cc (hunk 4 above)
Comment (tempdragon) [reply gcc/jit/libgccjit.cc]: Resolved by deleting  `recording::block::set_block`

**Comment by tempdragon** [comment]
Thanks for your advice and (more importantly) tolerance.

File: gcc/jit/jit-recording.cc (hunk 1 above)
Comment (tempdragon) [reply gcc/jit/jit-recording.cc]: Resolved by deleting  `recording::block::set_block`. Which applies to all other`set_loc`.

File: gcc/jit/jit-recording.cc (hunk 2 above)
Comment (antoyo) [reply gcc/jit/jit-recording.cc]: > Maybe I should simply remove all other code.

I guess it depends if you're sure we'll use all the methods you added.

Comment (tempdragon) [reply gcc/jit/jit-recording.cc]: No. Even in the long round for set location of a block.

But there is a chance where I or someone else will use the other remaining set locations, except the rvalue one in use.

In that case, should I keep everything else not currently in use or simply commit ONLY the `rvalue::set_loc`?

Comment (tempdragon) [reply gcc/jit/jit-recording.cc]: Maybe, if someone plans to separate debuginfo from code generation, as is done in llvm.
But of course, we can still leave it as it is, as I explained in my cg_gcc commit.

**Comment by tempdragon** [comment]
If necessary, I can also remove the other functions not in use, leaving only the `rvalue::set_loc` and the corresponding `gcc_jit_rvalue_set_location`.

File: gcc/jit/jit-recording.cc (hunk 2 above)
Comment (antoyo-light) [reply gcc/jit/jit-recording.cc]: I'm still not sure I understand after reading your comment [here](https://github.com/rust-lang/rustc_codegen_gcc/pull/455#issuecomment-1962828082), so let me know if this is correct: these methods will be used when rustc_codegen_gcc will not set the debuginfo in `builder.rs`, but will set them separately like is done for `rustc_codegen_llvm`?

Comment (antoyo) [reply gcc/jit/jit-recording.cc]: I'm still not sure I understand after reading your comment [here](https://github.com/rust-lang/rustc_codegen_gcc/pull/455#issuecomment-1962828082), so let me know if this is correct: these methods will be used when rustc_codegen_gcc will not set the debuginfo in builder.rs, but will set them separately like is done for rustc_codegen_llvm?

Comment (tempdragon) [reply gcc/jit/jit-recording.cc]: Yes.

**Review by antoyo — APPROVED** [review]
Thanks a lot for your contribution!

**Pull request merged**